<a href="https://colab.research.google.com/github/MarioROT/IHLT-MAI/blob/main/Session2_MarioRosas_AlamLopez.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Lab session 2 (Document structure) - ILTH

**Students:**
- Mario Rosas
- Alam Lopez

**Lab Professor:** Salvador Medina Herrera

## Paraphrases Template

In [1]:
%%shell
git clone https://github.com/mariorot/IHLT-MAI.git
cd 'IHLT-MAI'
mv 'complementary_material' /content/
mv scripts /content/

Cloning into 'IHLT-MAI'...
remote: Enumerating objects: 61, done.
remote: Counting objects: 100% (61/61), done.
remote: Compressing objects: 100% (55/55), done.
remote: Total 61 (delta 22), reused 32 (delta 5), pack-reused 0
Receiving objects: 100% (61/61), 134.00 KiB | 3.12 MiB/s, done.
Resolving deltas: 100% (22/22), done.


In [7]:
from scripts.compute_metrics import ComputeMetrics
from scripts.text_preprocessing import TextPreprocessing
import pandas as pd
from scipy.stats import pearsonr

In [3]:
dt = pd.read_csv('complementary_material/test-gold/STS.input.SMTeuroparl.txt',sep='\t',header=None)
dt['gs'] = pd.read_csv('complementary_material/test-gold/STS.gs.SMTeuroparl.txt',sep='\t',header=None)

## TODO
1. Compute the Jaccard similarity between every paraphrase and add a column *jaccard* to *dt* variable.
2. Compute the pearson correlation as: <br>
```
from scipy.stats import pearsonr
pearsonr(dt['gs'], dt['jaccard'])[0]
```

The code for the tokenization is at: https://github.com/MarioROT/IHLT-MAI/blob/main/scripts/text_preprocessing.py

The code for jaccard distance computation is at: https://github.com/MarioROT/IHLT-MAI/blob/main/scripts/compute_metrics.py

In [4]:
tp = TextPreprocessing()
dt[2] = tp.tokenize_data(list(dt[0]),'nltk')
dt[3] = tp.tokenize_data(list(dt[1]),'nltk')
cm_nltk = ComputeMetrics(dt[[2,3]].to_numpy(), ['jaccard'], 1)
dt['jaccard_nltk'] = cm_nltk.do()['jaccard']

dt[4] = tp.tokenize_data(list(dt[0]),'spacy')
dt[5] = tp.tokenize_data(list(dt[1]),'spacy')

cm_spacy = ComputeMetrics(dt[[4,5]].to_numpy(), ['jaccard'], 1)
dt['jaccard_spacy'] = cm_spacy.do()['jaccard']

Computing jaccard...
Computing jaccard...


In [5]:
dt[['gs','jaccard_nltk','jaccard_spacy']]

gs  jaccard_nltk  jaccard_spacy
0    4.500      0.653846       0.653846
1    5.000      0.214286       0.214286
2    4.250      0.608696       0.608696
3    4.500      0.454545       0.400000
4    5.000      0.000000       0.000000
..     ...           ...            ...
454  5.000      0.450000       0.450000
455  4.750      0.642857       0.714286
456  5.000      0.600000       0.600000
457  4.000      0.681818       0.681818
458  3.833      0.500000       0.500000

[459 rows x 3 columns]

In [8]:
p_nltk=pearsonr(dt['gs'], 1-dt['jaccard_nltk'])[0] # Calculating the pearson correlation between GS results and 1-Jaccard calculated dats
p_spacy=pearsonr(dt['gs'], 1-dt['jaccard_spacy'])[0]
print(f'NLTK {p_nltk} --- spaCy {p_spacy}')

NLTK 0.45049771693186846 --- spaCy 0.4608504322255335


# Conclusion
The use of different tokenizers (NLTK and spaCy) can impact the results when comparing sentence pairs, as they may perform differently depending on the input text data. Jaccard similarity proved to be an effective metric for comparing the similarity of sentence pairs after word tokenization. With the inversion of this metric we got a distance to measure the dissimilarity between sets of tokens, making it suitable for the task we were trying to solve. The Pearson correlation coefficient provided us a quantitative measure of how well our results align with the gold-standard, i.e. effectiveness.

The results indicate that spaCy slightly outperformed NLTK in the task of sentence pair comparison using Jaccard distance.The Pearson correlation coefficient values obtained for **NLTK (0.4505)** and **spaCy (0.4609)** suggest a moderate positive correlation between the computed distances and the benchmark (gold-standard) distances.

Our best result could be possibly improved by experimenting with different tokenization strategies or exploring other NLP techniques to enhance the performance.


